contents/Documentation/FP concepts.playground/Pages/Functions vs Procedures.xcplaygroundpage/Contents.swift

```diff
@@ -4,11 +4,171 @@
+ */
+let memoizedFactorial: (Int) -> Int = memoize { factorial, x in
+    x == 0 ? 1 : x * factorial(x - 1)
+}
+
+let fact7 = memoizedFactorial(7)
+let fact4 = memoizedFactorial(4)
+let fact9 = memoizedFactorial(9)
+/*:
+ In this case, computing the factorial of 7 also saves any intermediate steps to compute its value, so when we ask for the factorial of 4, it is already in the lookup table. Finally, asking for factorial of 9 starts computing and stops when it needs to compute factorial of 7, which was already computed above.
+
+ ## Function composition
+
+ Referentially transparent functions are the building blocks for our programs, but we need some operation to combine them. The essential operation for functions is function composition. Bow provides functionality to compose two functions easily.
+ */
+func f1(_ x: Int) -> Int {
+    return 2 * x
+}
+
+func f2(_ x: Int) -> String {
+    return "\(x)"
+}
+
+let composed1 = compose(f2, f1)
+let composed2 = f2 <<< f1
+/*:
+ Both the `compose` function and the `<<<` operator receive two functions and provide a new function which behaves like applying both functions sequentially. `f2 <<< f1` is read as *f2 after f1*; that is, in the resulting function, `f1` is applied to the input and its output is then fed to `f2`.
+
+ In some cases, `compose` can be difficult to read right to left, or simply is not convenient to use. For those cases, Bow has utility functions that reverse the order of the arguments.
+ */
+let composed3 = andThen(f1, f2)
+let composed4 = f1 >>> f2
+/*:
+ Those calls are equivalent to the ones above, but with arguments reversed. `f1 >>> f2` is read as *apply f1 and then f2*, and the behavior is exactly the same as in the `<<<` operator.
+
+ Composition of functions is associative; i.e. the following functions are equivalent:
+ */
+func f3(_ x: String) -> String {
+    return String(x.reversed())
+}
+
+let associativity1 = (f3 <<< f2) <<< f1
+let associativity2 = f3 <<< (f2 <<< f1)
+/*:
+ Given three functions, it does not matter if we compose the first two, and then the third one, or the latter two and then the first one. There is also a function that, when composed with any other function, does nothing: the `id` (identity) function, which is included in Bow.
+ */
+let identity1 = f1 <<< id
+let identity2 = id <<< f1
+/*:
+ The functions above have the exact same behavior as `f1`.
+
+ ## Working with impure functions
+
+ Real world software rarely has functions as nicely written as the `add` function above. We usually have to deal with partial, non-deterministic and/or impure functions, since software has to deal with errors and perform side effects in order to be useful. Besides presenting issues breaking referential transparency, function composition cannot be done in this circumstances. How can we deal with such situations?
```

Comment:
```suggestion
 Real world software rarely has functions as nicely written as the `add` function above. We usually have to deal with partial, non-deterministic and/or impure functions, since software has to deal with errors and perform side effects in order to be useful. Besides presenting issues breaking referential transparency, function composition cannot be done in these circumstances. How can we deal with such situations?
```